File: src/main/java/org/funivan/lologa/tile/Tile.java
package org.funivan.lologa.tile;

import org.funivan.lologa.tile.Position.Position;
import org.funivan.lologa.tile.Position.PositionInterface;
import org.funivan.lologa.tile.Score.Score;
import org.funivan.lologa.tile.Score.ScoreInterface;

import java.awt.*;
import java.util.Objects;

public class Tile implements TileInterface {
    public static Tile DUMMY;

    static {
        DUMMY = new Tile(new Color(255, 255, 255), new Score(-100), new Position(-100, -100));
    }

    private final Color color;
    private ScoreInterface score;
    private PositionInterface position;

    public Tile(Color color, ScoreInterface score, PositionInterface position) {
        this.color = color;
        this.score = score;
        this.position = position;
    }

    public Tile(Color color, PositionInterface position) {
        this(color, new Score(1), position);
    }

    public Color color() {
        return this.color;
    }

    @Override
    public ScoreInterface score() {
        return this.score;
    }


    @Override
    public PositionInterface position() {
        return this.position;
    }

    @Override
    public boolean equals(Object o) {
        boolean result = false;
        if (this == o) {
            result = true;
        } else if (o instanceof TileInterface) {
            TileInterface target = (TileInterface) o;
            result = (
                target.color().equals(this.color()) && target.position().equals(this.position())
            );
        }
        return result;
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.color, this.score, this.position);
    }

    @Override
    public boolean same(TileInterface target) {
        return (target.score().value() == this.score().value()
            &&
            target.color().equals(this.color())
            &&
            target.position().equals(this.position())
        );
    }

    @Override
    public String toString() {
        return "Tile: " +
            "{Color{r" + this.color.getRed() + ",g" + this.color.getGreen() + ",b" + this.color.getBlue() + "}}"
            + "," + this.score + ""
            + "," + this.position;
    }
}

File: src/main/java/org/funivan/lologa/tile/Tile.kt
package org.funivan.lologa.tile

import org.funivan.lologa.tile.Position.Position
import org.funivan.lologa.tile.Position.PositionInterface
import org.funivan.lologa.tile.Score.Score
import org.funivan.lologa.tile.Score.ScoreInterface
import java.awt.Color
import java.util.*

class Tile(private val color: Color, private val score: ScoreInterface, private val position: PositionInterface) : TileInterface {

    constructor(color: Color, position: PositionInterface) : this(color, Score(1), position) {}

    constructor(score: Score, position: PositionInterface) : this(Color.WHITE, score, position) {}

    override fun color(): Color {
        return this.color
    }

    override fun score(): ScoreInterface {
        return this.score
    }


    override fun position(): PositionInterface {
        return this.position
    }

    override fun equals(o: Any?): Boolean {
        var result = false
        if (this === o) {
            result = true
        } else if (o is TileInterface) {
            val target = o as TileInterface?
            result = target!!.color() == this.color() && target.position() == this.position()
        }
        return result
    }

    override fun hashCode(): Int {
        return Objects.hash(this.color, this.score, this.position)
    }

    override fun same(target: TileInterface): Boolean {
        return (target.score().value() == this.score().value()
                &&
                target.color() == this.color()
                &&
                target.position() == this.position())
    }

    override fun toString(): String {
        return ("Tile: " +
                "{Color{r" + this.color.red + ",g" + this.color.green + ",b" + this.color.blue + "}}"
                + "," + this.score + ""
                + "," + this.position)
    }

    companion object {

        val DUMMY: TileInterface = Tile(Color(0, 0, 0), Score(-1), Position(-1, -1))
    }
}
